File: src/data_maki/katanuki.py
from . import utils
from .context import Context
from .models.answer import Direction, Op
from .patterns import get_pattern
from .utils import ReverseOperation
from .models.replay import MarkType, CellsMark, ExtraOpInfo, ReplayInfo


def add_ops(c: Context, p: int, x: int, y: int, s: Direction):
    pattern = get_pattern(p, c.patterns)

    # this function can't correct x&y confused by ud or lr
    if c.rv_op.has_reverse90:
        x, y = y, x

        if c.rv_op.has_reverse_up_down:
            y -= pattern.height - 1

            if s == Direction.LEFT:
                s = Direction.RIGHT
            elif s == Direction.RIGHT:
                s = Direction.LEFT

        if c.rv_op.has_reverse_left_right:
            x -= pattern.width - 1

            if s == Direction.UP:
                s = Direction.DOWN
            elif s == Direction.DOWN:
                s = Direction.UP

        if s == Direction.UP:
            s = Direction.LEFT
        elif s == Direction.DOWN:
            s = Direction.RIGHT
        elif s == Direction.LEFT:
            s = Direction.UP
        elif s == Direction.RIGHT:
            s = Direction.DOWN

    else:
        if c.rv_op.has_reverse_up_down:
            y -= pattern.height - 1

            if s == Direction.UP:
                s = Direction.DOWN
            elif s == Direction.DOWN:
                s = Direction.UP

        if c.rv_op.has_reverse_left_right:
            x -= pattern.width - 1

            if s == Direction.LEFT:
                s = Direction.RIGHT
            elif s == Direction.RIGHT:
                s = Direction.LEFT

    c.n += 1
    c.ops.append(Op(p, x, y, s))


def add_info(c: Context):
    gl_type = c.info_now.goalMark.type
    gl_idx = c.info_now.goalMark.index
    cr_type = c.info_now.currentMark.type
    cr_y = c.info_now.currentMark.index
    cr_x = c.info_now.currentMark.index2

    if c.rv_op.has_reverse90:
        if gl_type == MarkType.ROW:
            gl_type = MarkType.COLUMN
        else:
            gl_type = MarkType.ROW

        cr_x, cr_y = cr_y, cr_x

    if c.rv_op.has_reverse_up_down:
        if gl_type == MarkType.ROW:
            gl_idx = c.height - gl_idx - 1

        cr_y = c.height - cr_y - 1

    if c.rv_op.has_reverse_left_right:
        if gl_type == MarkType.COLUMN:
            gl_idx = c.width - gl_idx - 1

        cr_x = c.width - cr_x - 1

    if c.info_now.goalMark.type == MarkType.ROW:
        c.info.append(ExtraOpInfo(CellsMark(cr_type, cr_y, cr_x), CellsMark(gl_type, gl_idx, None), c.info_now.delta))
    else:
        c.info.append(ExtraOpInfo(CellsMark(cr_type, cr_y, cr_x), CellsMark(gl_type, gl_idx, None), None))


def katanuki_board(c: Context, p: int, x: int, y: int, s: Direction):
    pattern_ = get_pattern(p, c.patterns)
    pattern = utils.reverse(pattern_.cells, c.rv_op)

    if x + pattern_.width <= 0 or x >= c.width or y + pattern_.height <= 0 or y >= c.height:
        raise Exception("Nukigata can't pick any cells :(")

    # stripe -> reverse / border -> normal
    if s == Direction.UP or s == Direction.DOWN:
        b = utils.list_rv(c.board.current, ReverseOperation.Reverse90)
        bx = c.height
        by = c.width
        pw = pattern.height
        ph = pattern.width
        px = y
        py = x
        pattern = utils.list_rv(pattern, ReverseOperation.Reverse90)
    elif s == Direction.LEFT or s == Direction.RIGHT:
        b = c.board.current.copy()
        bx = c.width
        by = c.height
        pw = pattern.width
        ph = pattern.height
        px = x
        py = y
        # pattern: do nothing
    else:
        raise Exception("the direction is not exist! :(")

    """
    pattern = utils.list_rv(pattern, c.rv_uldr, c.rv_ud, c.rv_lr)
    """

    """
    before
    3 2 1
    | | |
    | | |
    V V V

    after
    <---1
    <---2
    <---3
    """

    for i in range(ph):
        ly = py + i
        # out of range
        if ly < 0:
            continue
        if ly >= by:
            break

        now_col: list[int] = b.get_row(ly)
        # nukigata de nuita cell
        picked = []

        for j in range(pw - 1, -1, -1):
            lx = px + j
            # out of range
            if lx >= bx:
                continue
            if lx < 0:
                break

            if pattern.get(i, j) == 0:
                continue

            # pick and remove the cell looked at
            picked.insert(0, now_col.pop(lx))

        if s == Direction.UP or s == Direction.LEFT:
            now_col += picked
        else:
            now_col = picked + now_col

        b.set_row(ly, now_col)

    if s == Direction.UP or s == Direction.DOWN:
        b = utils.list_rv(b, ReverseOperation.Reverse90)

    return b


def katanuki(c: Context, p: int, x: int, y: int, s: Direction):
    print([p, x, y, s])

    b = katanuki_board(c, p, x, y, s)

    if c.board.current != b:
        c.board.current = b

        add_ops(c, p, x, y, s)

        c.elems_now = utils.count_elements(c.board.current)

        utils.print_board(c.board.current)

        add_info(c)

    return

File: packages/algorithm/src/vbeam/katanuki.ts
import type { Op } from "@data-maki/schemas";
import { getPattern } from "../models/pattern";
import { type Context, DOWN, type Direction, LEFT, type Point, RIGHT, UP } from "../types";
import { type TwoDimensionalCells, dbgCells, multiReverse, reverseCells } from "../utils/arrays";
import { countElementsColumnWise } from "../utils/board";
import { dbg } from "../workers/log";

export const addOp = (c: Context, op: Op) => {
  let { p, x, y, s } = op;

  const pattern = getPattern(op.p, c.patterns);

  if (c.rvOp.hasReverse90) {
    const tmp = x;
    x = y;
    y = tmp;

    if (c.rvOp.hasReverseUpDown) {
      y = c.width - y - 1;
      y -= pattern.height - 1;

      if (s === LEFT) s = RIGHT;
      else if (s === RIGHT) s = LEFT;
    }

    if (c.rvOp.hasReverseLeftRight) {
      x = c.height - x - 1;
      x -= pattern.width - 1;

      if (s === UP) s = DOWN;
      else if (s === DOWN) s = UP;
    }

    if (s === UP) s = LEFT;
    else if (s === DOWN) s = RIGHT;
    else if (s === LEFT) s = UP;
    else if (s === RIGHT) s = DOWN;
  } else {
    if (c.rvOp.hasReverseUpDown) {
      y = c.height - y - 1;
      y -= pattern.height - 1;

      if (s === UP) s = DOWN;
      else if (s === DOWN) s = UP;
    }

    if (c.rvOp.hasReverseLeftRight) {
      x = c.width - x - 1;
      x -= pattern.width - 1;

      if (s === LEFT) s = RIGHT;
      else if (s === RIGHT) s = LEFT;
    }
  }

  c.n += 1;
  c.ops.push({ p, x, y, s });
};

const generatePatternData = (
  dir: Direction,
  b: TwoDimensionalCells,
  pattern: TwoDimensionalCells,
  w: number,
  h: number,
  x: number,
  y: number,
) =>
  dir === UP || dir === DOWN
    ? {
        b: reverseCells(b, "reverse-90"),
        bx: h,
        by: w,
        pw: pattern.height,
        ph: pattern.width,
        pp: {
          x: y,
          y: x,
        },
        pattern: reverseCells(pattern, "reverse-90"),
      }
    : /* dir === LEFT || dir === RIGHT */ {
        b: b.clone(),
        bx: w,
        by: h,
        pw: pattern.width,
        ph: pattern.height,
        pp: {
          x,
          y,
        },
        pattern,
      };

export const katanukiBoard = (c: Context, p: number, x: number, y: number, s: Direction, _b: TwoDimensionalCells) => {
  const pattern_ = getPattern(p, c.patterns);

  dbg(c.worker, "katanuki", { p: pattern_.p, x, y, s });

  if (x + pattern_.width <= 0 || x >= c.width || y + pattern_.height <= 0 || y >= c.height) {
    throw new Error(
      `Cannot pick any cells: out of range (x: ${c.width} > [${x}, ${x + pattern_.width}), y: ${c.height} > [${y}, ${y + pattern_.height})`,
    );
  }

  // Looking point on the board
  const l: Point = { x: 0, y: 0 };

  // Stripe -> Reverse / Border -> Normal
  let { b, bx, by, pw, ph, pp, pattern } = generatePatternData(
    s,
    _b,
    multiReverse(pattern_.cells, c.rvOp),
    c.width,
    c.height,
    x,
    y,
  );

  for (const i of Array(ph).keys()) {
    l.y = pp.y + i;

    // Out of range
    if (l.y < 0) continue;
    if (l.y >= by) break;

    const currentRow = b.getRow(l.y);
    const picked: number[] = [];

    for (let j = pw - 1; j > -1; j--) {
      l.x = pp.x + j;

      if (l.x >= bx) continue;
      if (l.x < 0) break;

      if (pattern.get(i, j) === 0) continue;

      picked.unshift(currentRow.splice(l.x, 1)[0]);
    }

    if (s === UP || s === LEFT) {
      currentRow.push(...picked);
    } else {
      currentRow.unshift(...picked);
    }

    b.setRow(l.y, currentRow);
  }

  if (s === UP || s === DOWN) {
    b = reverseCells(b, "reverse-90");
  }

  return b;
};

export const katanuki = (c: Context, p: number, x: number, y: number, s: Direction) => {
  const b = katanukiBoard(c, p, x, y, s, c.board);

  if (!c.board.equals(b)) {
    c.board = b;
    c.currentElementCounts = countElementsColumnWise(c.board, c.worker);

    addOp(c, { p, x, y, s });
    dbgCells(c.board, c.worker);
  } else {
    dbg(c.worker, c.board, b);
  }
};
